hello world blog post (#104)
spinkube/documentation: content/en/blog/news/first-post/index.md

---
date: 2024-03-13
title: Introducing SpinKube
linkTitle: Introducing SpinKube
description: >
  Bringing the next wave of hyper-efficient serverless to Kubernetes.
author: The SpinKube Team ([@SpinKube](https://mastodon.social/@SpinKube))
resources:
  - src: "**.{png,jpg}"
    title: "Image #:counter"
---
Today we're introducing SpinKube - an open source platform for efficiently 

> running Spin-based WebAssembly applications on Kubernetes. Built with love by 

Reviewer: ```suggestion
running Spin-based WebAssembly (Wasm) applications on Kubernetes. Built with love by 
```

folks from Microsoft, SUSE, LiquidReply, and Fermyon.

> SpinKube combines the application lifecycle management of the [Spin Operator][spin-operator], the efficiency of the [containerd-shim-spin][containerd-shim-spin], and the node lifecycle management of the forthcoming [runtime-class-manager][runtime-class-manager] (formerly [KWasm][kwasm]) to provide an excellent developer and operator experience alongside excellent density and scaling characteristics regardless of processor architecture or deployment environment.

Reviewer: ```suggestion
SpinKube combines the application lifecycle management of the [Spin Operator](https://github.com/spinkube/spin-operator), the efficiency of the [containerd-shim-spin](https://github.com/spinkube/containerd-shim-spin), and the node lifecycle management of the forthcoming [runtime-class-manager](https://github.com/spinkube/runtime-class-manager) (formerly [KWasm](https://kwasm.sh/)) to provide an excellent developer and operator experience alongside excellent density and scaling characteristics regardless of processor architecture or deployment environment.
```

>

> > TL;DR? Check out the [quickstart][quickstart] to learn how to deploy

Reviewer: ```suggestion
> TL;DR? Check out the [quickstart](https://www.spinkube.dev/docs/spin-operator/quickstart/) to learn how to deploy
```

Reviewer 2: This is a reference-style link and is considered valid markdown.

> > WebAssembly to Kubernetes today.

Reviewer: ```suggestion
> Wasm to Kubernetes today.
```

> 

## Why Serverless?

Containers and Kubernetes revolutionized the field of software development and operations. A unified packaging and dependency management system that was flexible enough to allow most applications to become portable - to run the same versions locally as you did in production - was a desperately needed iteration from the Virtual Machines that came before.

But while containers give us a relatively light weight abstraction layer, they come with a variety of issues:

- *Change management* - patching system dependencies (such as OpenSSL) need to happen on every image independently, which can become difficult to manage at large scale. 
- *Rightsizing workloads* - managing access to shared resources like CPU and memory correctly is difficult, which often leads to over-provisioning resources (over introducing downtime) resulting in low utilization of host resources. 
- *Size* - containers are often bigger than needed (impacting important metrics like startup time), and often include extraneous system dependencies by default which are not required to run the application

The desire to fix many of these issues is what led to the "first" generation of Serverless.

>

> > Borrowing ideas from CGI, Serverless apps are not written as persistent servers. Instead Server-less code is code that responds to events (such as HTTP requests), but does not run as a daemon process listening on a socket. The networking part of serverless architectures is relegated to the underlying infrastructure.

Author: Going to reject this one because I want to avoid conflating `component` - but in this case it _is_ lower level (you only shipping your business logic is a good thing 😅 )

> 

> First generation serverless runtimes (such as AWS Lambda, Google Cloud Functions, Azure Functions, and their Kubernetes counterparts OpenWhisk and KNative) are all based on the principle of running a VM or container per function. While this afforded flexibility for application developers, complexity was introduced to platform engineers as neither compute type is designed to start quickly. Platform engineers operating these serverless platforms became responsible for an elaborate dance of pre-warming compute capacity and loading a workload just-in-time, making for a difficult tradeoff in cold start performance and cost. The result? A slow and inefficient first generation of Serverless.

Author: I think this is caveat-ed _enough_ by the container/vm-based distinction - although it definitely has the chance to stale-over-time, I think being accurate-at-the-time-of-publication is ok in context.

To illustrate this point, below are all the steps required to start a Kubernetes Pod:

| Phase | Description |
| --- | --- |
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |
| Image pull (cacheable) | The node pulls the container images from a remote repository to a local cache when needed |
| Image mount | The node prepares the container images and mounts them |
| Container start | The node configures and starts the Pod's containers and sidecars |
| Application load | Each container loads application code and libraries, and initializes application memory. |
| Application initialization | Each container performs application-specific init code (entry-point, setting up telemetry, validating connections to external dependencies, etc.) |
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

> Circumventing this process, and the amount of duplication that occurs on every replica even if they are running on the same physical hardware, is where the [Spin][spin] WebAssembly Runtime and SpinKube shines.

Reviewer: ```suggestion
Circumventing this process, and the amount of duplication that occurs on every replica even if they are running on the same physical hardware, is where the [Spin][https://github.com/fermyon/spin] Wasm Runtime and SpinKube shine.
```

## Why SpinKube?

SpinKube solves or mitigates many of the issues commonly associated with deploying, scaling, and operating Serverless applications.

> The foundational work that underpins the majority of this comes from removing the Container. Spin Applications are primarily distributed as OCI Artifacts, rather than traditional container images, meaning that you only ship your compiled application and its assets - without system dependencies.

Author: We still use the runtime (containerd) - so I was using `Container` here to refer to quite specifically the whole "layered filesystem that must be mounted and chroot-ed before executing a thing".

I thought images were considered OCI *a*rtifacts, but not OCI *A*rtifacts? 😅

> The foundational work that underpins the majority of this comes from removing the Container. Spin Applications are primarily distributed as OCI Artifacts, rather than traditional container images, meaning that you only ship your compiled application and its assets - without system dependencies.

Reviewer: A suggestion is to use `Wasm` instead of `WebAssembly` if/when this inline comment is committed.
(Spelling out WebAssembly followed by Wasm in parentheses at the start of the article, then using Wasm alone in subsequent references.)

> The foundational work that underpins the majority of this comes from removing the Container. Spin Applications are primarily distributed as OCI Artifacts, rather than traditional container images, meaning that you only ship your compiled application and its assets - without system dependencies.

Reviewer 2: Are we still using the container runtime provided by containerd, or are we using containerd to act as the interface between the Container Runtime Interface (CRI) and the actual runtime (Spin)? I thought a containerd shim uses containerd to interface with the CRI, but swapped the container runtime with something else (in this case, the Spin runtime).

> I thought images were considered OCI artifacts, but not OCI Artifacts?

Container images, singularity images, helm charts, binary blobs, and WebAssembly components are all considered OCI artifacts.

> The foundational work that underpins the majority of this comes from removing the Container. Spin Applications are primarily distributed as OCI Artifacts, rather than traditional container images, meaning that you only ship your compiled application and its assets - without system dependencies.

Reviewer 2: I feel like I might be making a mountain out of a molehill here... Just thought I’d throw out some edits to help clarify some terminology to new readers. Feel free to accept or reject them.

> The foundational work that underpins the majority of this comes from removing the Container. Spin Applications are primarily distributed as OCI Artifacts, rather than traditional container images, meaning that you only ship your compiled application and its assets - without system dependencies.

Reviewer 3: >I thought images were considered OCI artifacts, but not OCI Artifacts?

For context, Spin applications are indeed distributed as _OCI Artifacts_ (capital A, from https://github.com/opencontainers/artifacts, which made its way to the spec).
So the initial sentence sounds good to me.

> To execute those applications we have a [runwasi](https://github.com/containerd/runwasi) based [containerd-shim-spin](https://github.com/spinkube/containerd-shim-spin/) that takes a Spin app, [pre-compiles it for the specific architecture](https://github.com/spinkube/containerd-shim-spin/pull/32) (caching the result in the containerd content store), and is then ready to service requests with [sub-millisecond startup times](https://fermyon.github.io/spin-benchmarks/criterion/reports/spin-executor_sleep-1ms/index.html) - no matter how long your application has been sitting idle.

Author: hyphenating is correct here (so I'm going to reject this change specifically)

> This also has the benefit of moving the core security patching of web servers, queue listeners, and other application boundaries to the host, rather than image artifacts themselves. But what if you mostly rely on images provided by your cloud-provider? Or deploy to environments that are not easily reproducible?

Reviewer: Suggest using `cloud provider` instead of `cloud-provider` (not a compound adjective).

> This is where [runtime-class-manager](https://github.com/spinkube/runtime-class-manager) (coming soon, formerly KWasm) comes into play - a Production Ready and Kubernetes-native way to manage WebAssembly runtimes on Kubernetes Hosts. With runtime-class-manager you can ensure that containerd-shim-spin is installed on your hosts, manage version lifecycles, and security patches - all from the Kubernetes API.

Reviewer: Same suggestion here:
A suggestion is to use Wasm instead of WebAssembly if/when this inline comment is committed.
(Spelling out WebAssembly followed by Wasm in parentheses at the start of the article, then using Wasm alone in subsequent references.)

> By combining these technologies, scaling a WebAssembly application looks more like this:

Reviewer: ```suggestion
By combining these technologies, scaling a Wasm application looks more like this:
```

| Phase | Description |
| --- | --- |
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |
| Image pull (cacheable) | The node pulls the needed container images from a remote repository to a local cache when needed |

> | WASM Load (cachable) | The shim loads the webassembly and prepares it for execution on the node’s CPU architecture |

Reviewer: ```suggestion
| Wasm Load (cachable) | The shim loads the Wasm and prepares it for execution on the node’s CPU architecture |
```

| Application start | The shim starts listening on the configured port and is ready to serve |
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

Or in practice:

<video src="spinkube-scaling.mp4" width="80%" controls></video>

## Deploying applications

This leaves one major area that can be painful when deploying applications to Kubernetes today: the developer and deployment experience.

> This is where [Spin][spin], [the `spin kube` plugin](https://github.com/spinkube/spin-plugin-kube), and [`spin-operator`][spin-operator] start to shine.

Reviewer: ```suggestion
This is where [Spin](https://github.com/fermyon/spin), [the `spin kube` plugin](https://github.com/spinkube/spin-plugin-kube), and [`spin-operator`](https://github.com/spinkube/spin-operator) start to shine.
```

Reviewer 2: These are reference-style links and are considered valid markdown

>
> TL;DR? Check out the [quickstart][quickstart] to learn how to deploy
> WebAssembly to Kubernetes today.
> 

## Why Serverless?

Containers and Kubernetes revolutionized the field of software development and operations. A unified packaging and dependency management system that was flexible enough to allow most applications to become portable - to run the same versions locally as you did in production - was a desperately needed iteration from the Virtual Machines that came before.

But while containers give us a relatively light weight abstraction layer, they come with a variety of issues:

- *Change management* - patching system dependencies (such as OpenSSL) need to happen on every image independently, which can become difficult to manage at large scale. 
- *Rightsizing workloads* - managing access to shared resources like CPU and memory correctly is difficult, which often leads to over-provisioning resources (over introducing downtime) resulting in low utilization of host resources. 

> - *Size* - containers are often bigger than needed (impacting important metrics like startup time), and often include extraneous system dependencies by default which are not required to run the application

Reviewer: ```suggestion
- *Size* - containers are often bigger than needed (impacting important metrics like startup time), and often include extraneous system dependencies by default which are not required to run the application.
```

>
> Borrowing ideas from CGI, Serverless apps are not written as persistent servers. Instead Server-less code is code that responds to events (such as HTTP requests), but does not run as a daemon process listening on a socket. The networking part of serverless architectures is relegated to the underlying infrastructure.
> 

First generation serverless runtimes (such as AWS Lambda, Google Cloud Functions, Azure Functions, and their Kubernetes counterparts OpenWhisk and KNative) are all based on the principle of running a VM or container per function. While this afforded flexibility for application developers, complexity was introduced to platform engineers as neither compute type is designed to start quickly. Platform engineers operating these serverless platforms became responsible for an elaborate dance of pre-warming compute capacity and loading a workload just-in-time, making for a difficult tradeoff in cold start performance and cost. The result? A slow and inefficient first generation of Serverless.

To illustrate this point, below are all the steps required to start a Kubernetes Pod:

| Phase | Description |
| --- | --- |

> | Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |

Reviewer: ```suggestion
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod. |
```

> | Image pull (cacheable) | The node pulls the container images from a remote repository to a local cache when needed |

Reviewer: ```suggestion
| Image pull (cacheable) | The node pulls the container images from a remote repository to a local cache when needed. |
```

> | Image mount | The node prepares the container images and mounts them |

Reviewer: ```suggestion
| Image mount | The node prepares the container images and mounts them. |
```

> | Container start | The node configures and starts the Pod's containers and sidecars |

Reviewer: ```suggestion
| Container start | The node configures and starts the Pod's containers and sidecars. |
```

| Application load | Each container loads application code and libraries, and initializes application memory. |

> | Application initialization | Each container performs application-specific init code (entry-point, setting up telemetry, validating connections to external dependencies, etc.) |

Reviewer: ```suggestion
| Application initialization | Each container performs application-specific init code (entry-point, setting up telemetry, validating connections to external dependencies, etc.). |
```

> | Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |

Reviewer: ```suggestion
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready. |
```

> | Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

Reviewer: ```suggestion
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic. |
```

This also has the benefit of moving the core security patching of web servers, queue listeners, and other application boundaries to the host, rather than image artifacts themselves.

But what if you mostly rely on images provided by your cloud-provider? Or deploy to environments that are not easily reproducible?

This is where [runtime-class-manager][runtime-class-manager] (coming soon, formerly [KWasm][kwasm]) comes into play - a Production Ready and Kubernetes-native way to manage WebAssembly runtimes on Kubernetes Hosts. With runtime-class-manager you can ensure that containerd-shim-spin is installed on your hosts, manage version lifecycles, and security patches - all from the Kubernetes API.

By combining these technologies, scaling a WebAssembly application looks more like this:

| Phase | Description |
| --- | --- |

> | Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |

Reviewer: ```suggestion
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod. |
```

> | Image pull (cacheable) | The node pulls the needed container images from a remote repository to a local cache when needed |

Reviewer: ```suggestion
| Image pull (cacheable) | The node pulls the needed container images from a remote repository to a local cache when needed. |
```

| WASM Load (cachable) | The shim loads the webassembly and prepares it for execution on the node’s CPU architecture |

> | Application start | The shim starts listening on the configured port and is ready to serve |

Reviewer: ```suggestion
| Application start | The shim starts listening on the configured port and is ready to serve. |
```

> | Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |

Reviewer: ```suggestion
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready. |
```

> | Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

Reviewer: ```suggestion
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic. |
```

![Architecture diagram of the SpinKube project](spinkube-diagram.png)

The spin-operator makes managing serverless applications easy - you provide the
image and bindings to secrets, and the operator handles realizing that
configuration as regular Kubernetes objects.

Combining the operator with the `spin kube` plugin gives you superpowers, as it
streamlines the process of generating Kubernetes YAML for your application and
gives you a powerful starting point for applications.

This is all it takes to get a HTTP application running on Kubernetes (after
installing the containerd shim and the operator):

```bash
# Create a new Spin App
spin new -t http-rust --accept-defaults spin-kube-app
cd spin-kube-app

# Build the Spin App
spin build

# Push the Spin App to an OCI registry
export IMAGE_NAME=ttl.sh/spin-app-$(uuidgen):1h
spin registry push $IMAGE_NAME

# Scaffold Kubernetes manifests
spin kube scaffold -f $IMAGE_NAME > app.yaml

# Deploy to Kubernetes
kubectl apply -f app.yaml
```

If you want to try things out yourself, the easiest way is to follow the [quickstart guide][quickstart] - we can’t wait to see what you build.

> 

Reviewer: ```suggestion

## References

```

Reviewer 2: Not necessary. The links provided below are not rendered. See below comments

> [spin-operator]: https://github.com/spinkube/spin-operator

Reviewer: Please note: The double spaces at the end of each line (in this reference section) must stay so that each individual reference (URL) is on its own new line.

![Screenshot 2024-03-13 at 11 09 14](https://github.com/spinkube/documentation/assets/9831342/a295c11b-c421-4ab2-84ee-087c3fe7fa6c)

> [spin-operator]: https://github.com/spinkube/spin-operator

Author: these aren't supposed to show on the page - they're markdown link references, not content 😅

> [containerd-shim-spin]: https://github.com/spinkube/containerd-shim-spin

Reviewer 2: The original format is correct. It is known as a reference-style link.

See https://daringfireball.net/projects/markdown/syntax#link for more information.

> [runtime-class-manager]: https://github.com/spinkube/runtime-class-manager

Reviewer 2: The original format is correct. It is known as a reference-style link.

See https://daringfireball.net/projects/markdown/syntax#link for more information.

> [spin]: https://github.com/fermyon/spin

Reviewer 2: The original format is correct. It is known as a reference-style link.

See https://daringfireball.net/projects/markdown/syntax#link for more information.

> [quickstart]: https://www.spinkube.dev/docs/spin-operator/quickstart/

Reviewer 2: The original format is correct. It is known as a reference-style link.

See https://daringfireball.net/projects/markdown/syntax#link for more information.

But while containers give us a relatively light weight abstraction layer, they come with a variety of issues:

- patching system dependencies like OpenSSL needs to happen in every image independently
- managing access to shared resources like CPU/memory/disk is hard, and leads to over-provisioning resources which results in low-density and wasteful compute
- containers are often much bigger than they need to be - including far more system dependencies than are required to run an application.

The desire to fix many of these issues is what led to the "first" generation of Serverless.

>
> Borrowing ideas from CGI, Serverless apps are not written as persistent servers. Instead Server-less code is code that responds to events (such as HTTP requests), but does not run as a daemon process listening on a socket. The networking part of serverless architectures is relegated to the underlying infrastructure.
> 

> First generation serverless runtimes (such as AWS Lambda, Google Cloud Functions, Azure Functions, and their Kubernetes counterparts OpenWhisk and KNative) are all based on the principle of running a VM or container per function. While this is very flexible, neither compute type is designed to start quickly, an elaborate dance of pre-warming compute capacity and loading a workload just-in-time made the first generation of Serverless slow and inefficient.

Reviewer 4: Added a sentence or two to clarify what we mean by "flexibility" (assuming it's app dev model, but feel free to correct)

```suggestion
First generation serverless runtimes (such as AWS Lambda, Google Cloud Functions, Azure Functions, and their Kubernetes counterparts OpenWhisk and KNative) are all based on the principle of running a VM or container per function. While this afforded flexibility for application developers, complexity was introduced to platform engineers as neither compute type is designed to start quickly. Platform engineers operating these serverless platforms became responsible for an elaborate dance of pre-warming compute capacity and loading a workload just-in-time, making for a difficult tradeoff in cold start performance and cost. The result? A slow and inefficient first generation of Serverless.
```

> See the following that describes the process of starting a Kubernetes Pod:

Reviewer 4: ```suggestion
To illustrate this point, below are all the steps required to start a Kubernetes Pod:
```

| Phase | Description |
| --- | --- |
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |
| Image pull (cacheable) | The node pulls the container images from a remote repository to a local cache when needed |
| Image mount | The node prepares the container images and mounts them |
| Container start | The node configures and starts the Pod's containers and sidecars |
| Application load | Each container loads application code and libraries, and initializes application memory. |
| Application initialization | Each container performs application-specific init code (entry-point, setting up telemetry, validating connections to external dependencies, etc.) |
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

> This process, and the amount of duplication that occurs on every replica, even if they are running on the same physical hardware, is where the [Spin][spin] WebAssembly Runtime and SpinKube shines.

Reviewer 4: ```suggestion
Circumventing this process, and the amount of duplication that occurs on every replica even if they are running on the same physical hardware, is where the [Spin][spin] WebAssembly Runtime and SpinKube shines.
```

| Phase | Description |
| --- | --- |
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |
| Image pull (cacheable) | The node pulls the needed container images from a remote repository to a local cache when needed |
| WASM Load (cachable) | The shim loads the webassembly and prepares it for execution on the node’s CPU architecture |
| Application start | The shim starts listening on the configured port and is ready to serve |
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

Or in practice:

<video src="spinkube-scaling.mp4" width="80%" controls></video>

## Deploying applications

This leaves one major area that can be painful when deploying applications to Kubernetes today: the developer and deployment experience.

This is where [Spin][spin], [the `spin kube` plugin](https://github.com/spinkube/spin-plugin-kube), and [`spin-operator`][spin-operator] come into play.

![](spinkube-diagram.png)

> The spin-operator makes managing serverless applications easy - you provide the

Reviewer 4: Lil markdown issue here through line 101

Author: Can't tell what the issue would be? (from either github's preview rendering or hugo's output 😅 )

Reviewer 2: Possibly due to the lack of an alt tag on line 90?

> This is where [Spin][spin], [the `spin kube` plugin](https://github.com/spinkube/spin-plugin-kube), and [`spin-operator`][spin-operator] come into play.

Reviewer 5: nit: "come into play" used also 20 lines above when talking about the `runtime-class-manager`. Maybe we can come up with an alternative wording?

Author: done!

image and bindings to secrets, and the operator handles realizing that
configuration as regular Kubernetes objects.

Combining the operator with the `spin kube` plugin gives you superpowers, as it
streamlines the process of generating Kubernetes YAML for your application and
gives you a powerful starting point for applications.

This is all it takes to get a HTTP application running on Kubernetes (after
installing the containerd shim and the operator):

```bash
spin new -t http-rust spin-kube-app --accept-defaults
cd spin-kube-app
spin build
export IMAGE_NAME=ttl.sh/spin-app-$(uuidgen):1h
spin registry push $IMAGE_NAME
spin kube scaffold -f $IMAGE_NAME > app.yaml

> kubectl apply -f app.yaml

Reviewer 5: What about adding some inline comments like this

```bash
# Create a new Spin App
spin new -t http-rust --accept-defaults spin-kube-app
cd spin-kube-app

# Build the Spin App
spin build

# Push the Spin App to an OCI compliant registry
export IMAGE_NAME=ttl.sh/spin-app-$(uuidgen):1h
spin registry push $IMAGE_NAME

# Scaffold Kubernetes manifests
spin kube scaffold -f $IMAGE_NAME > app.yaml

# Deploy to Kubernetes
kubectl apply -f app.yaml
```

Author: done!

title: Introducing SpinKube
linkTitle: Introducing SpinKube
description: >
  Bringing the next wave of hyper-efficient serverless to Kubernetes.
author: The SpinKube Team ([@SpinKube](https://mastodon.social/@SpinKube))
resources:
  - src: "**.{png,jpg}"
    title: "Image #:counter"
---
Today we're introducing SpinKube - an open source platform for efficiently 
running Spin-based WebAssembly applications on Kubernetes. Built with love by 
folks from Microsoft, SUSE, LiquidReply, and Fermyon.

> SpinKube combines the application lifecycle management of the [Spin Operator][spin-operator], the efficiency of the [containerd-shim-spin][containerd-shim-spin], and the node lifecycle management of the forthcoming [runtime-class-manager][runtime-class-manager] (formerly KWasm) to provide an excellent developer and operator experience alongside excellent density and scaling characteristics regardless of processor architecture or deployment environment.

Reviewer 3: nit: link to kwasm.sh?

| Application load | Each container loads application code and libraries, and initializes application memory. |
| Application initialization | Each container performs application-specific init code (entry-point, setting up telemetry, validating connections to external dependencies, etc.) |
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

Circumventing this process, and the amount of duplication that occurs on every replica even if they are running on the same physical hardware, is where the [Spin][spin] WebAssembly Runtime and SpinKube shines.

## Why SpinKube?

SpinKube solves or mitigates many of the issues commonly associated with deploying, scaling, and operating Serverless applications.

The foundational work that underpins the majority of this comes from removing the Container. Spin Applications are primarily distributed as OCI Artifacts, rather than traditional container images, meaning that you only ship your compiled application and its assets - without system dependencies.

> To execute those applications we have a [runwasi](https://github.com/containerd/runwasi) based [containerd-shim-spin](https://github.com/spinkube/containerd-shim-spin/) that takes a WebAssembly artifact, [pre-compiles it for the specific architecture](https://github.com/spinkube/containerd-shim-spin/pull/32) (caching the result in the containerd content store), and is then ready to service requests with [sub-millisecond startup times](https://fermyon.github.io/spin-benchmarks/criterion/reports/spin-executor_sleep-1ms/index.html) - no matter how long your application has been sitting idle.

Reviewer 3: nit: technically it takes an entire Spin application, which can have several Wasm components; which you could refer to as a "WebAssembly artifact" if you want..

By combining these technologies, scaling a WebAssembly application looks more like this:

| Phase | Description |
| --- | --- |
| Kube init | Kubernetes schedules a Pod to a node, the node then prepares to run the Pod |
| Image pull (cacheable) | The node pulls the needed container images from a remote repository to a local cache when needed |
| WASM Load (cachable) | The shim loads the webassembly and prepares it for execution on the node’s CPU architecture |
| Application start | The shim starts listening on the configured port and is ready to serve |
| Pod ready | The Pod responds to the asynchronous readiness probe and signals that it is now ready |
| Service ready | The Pod gets added to the list of active endpoints for a Service and can start receiving traffic |

Or in practice:

<video src="spinkube-scaling.mp4" width="80%" controls></video>

## Deploying applications

This leaves one major area that can be painful when deploying applications to Kubernetes today: the developer and deployment experience.

This is where [Spin][spin], [the `spin kube` plugin](https://github.com/spinkube/spin-plugin-kube), and [`spin-operator`][spin-operator] start to shine.

> ![](spinkube-diagram.png)

Reviewer 2: I'd add an alt description for accessibility purposes:

```suggestion
![Architecture diagram of the SpinKube project](spinkube-diagram.png)
```

Author: Good spot (I totally forgot how to do that in markdown, didn't realize the text line just worked)

To execute those applications we have a [runwasi](https://github.com/containerd/runwasi) based [containerd-shim-spin](https://github.com/spinkube/containerd-shim-spin/) that takes a Spin app, [pre-compiles it for the specific architecture](https://github.com/spinkube/containerd-shim-spin/pull/32) (caching the result in the containerd content store), and is then ready to service requests with [sub-millisecond startup times](https://fermyon.github.io/spin-benchmarks/criterion/reports/spin-executor_sleep-1ms/index.html) - no matter how long your application has been sitting idle.

This also has the benefit of moving the core security patching of web servers, queue listeners, and other application boundaries to the host, rather than image artifacts themselves.

> But what if you mostly rely on images provided by your cloud-provider? Or deploy to environments that are not easily reproducible?

Reviewer: ```suggestion
But what if you mostly rely on images provided by your cloud provider? Or deploy to environments that are not easily reproducible?
```
